### PR TITLE
Various fixes

### LINKED ISSUE #18: If a pip install fails, it won't get retried.
If a `pip install -r requirements.txt` fails for whatever reason, it causes a cascade of problems:
- The next commands in the flow will all fail to complete.
- Future updates will not run the .venv commands, meaning it won't even self-correct.

### LINKED ISSUE #52: Split nginx config into staging and production?
I've been thinking about ways to solve the issue of needing to remove the SSL config for launch, only to then put it back once the DNS has propagated. One idea I had was to split the nginx config out into staging and production. With this approach we could configure the site with SSL and password protection during development, and have the non SSL production environment ready to go live from the start of the project, with LE just being run post launch. 

What do you think?

### LINKED ISSUE #54: Update the update command to update the app to the commit hash of Circle
Instead of just doing a `git pull`, we should be checking out the version that Circle just tested.

### LINKED ISSUE #55: Investigate zero downtime deploys

### LINKED ISSUE #57: nvm install failing
This is potentially an issue caused (at least in part) by the git clone pulling an empty master branch, instead of develop.

```
[178.62.105.236] out: => Profile not found. Tried  (as defined in $PROFILE), ~/.bashrc, ~/.bash_profile, ~/.zshrc, and ~/.profile.
[178.62.105.236] out: => Create one of them and run this script again
[178.62.105.236] out: => Create it (touch ) and run this script again
[178.62.105.236] out:    OR
[178.62.105.236] out: => Append the following lines to the correct file yourself:
[178.62.105.236] out:
[178.62.105.236] out: export NVM_DIR="/home/quotient/.nvm"
[178.62.105.236] out: [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
[178.62.105.236] out: => Close and reopen your terminal to start using nvm or run the following to use it now:
[178.62.105.236] out:
[178.62.105.236] out: export NVM_DIR="/home/quotient/.nvm"
[178.62.105.236] out: [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
[178.62.105.236] out:

[TASK] Install nvm ... done
[TASK] Activate nvm then install node and yarn ... [178.62.105.236] sudo: cd /var/www/quotient&& . ~quotient/.nvm/nvm.sh&& nvm install&& npm install -g yarn&& yarn&& yarn run build
[178.62.105.236] out: No .nvmrc file found
[178.62.105.236] out:
[178.62.105.236] out: Node Version Manager
[178.62.105.236] out:
[178.62.105.236] out: Note: <version> refers to any version-like string nvm understands. This includes:
[178.62.105.236] out:   - full or partial version numbers, starting with an optional "v" (0.10, v0.1.2, v1)
[178.62.105.236] out:   - default (built-in) aliases: node, stable, unstable, iojs, system
[178.62.105.236] out:   - custom aliases you define with `nvm alias foo`
[178.62.105.236] out:
[178.62.105.236] out:  Any options that produce colorized output should respect the `--no-colors` option.
[178.62.105.236] out:
[178.62.105.236] out: Usage:
[178.62.105.236] out:   nvm --help                                Show this message
[178.62.105.236] out:   nvm --version                             Print out the latest released version of nvm
[178.62.105.236] out:   nvm install [-s] <version>                Download and install a <version>, [-s] from source. Uses .nvmrc if available
[178.62.105.236] out:     --reinstall-packages-from=<version>     When installing, reinstall packages installed in <node|iojs|node version number>
[178.62.105.236] out:     --lts                                   When installing, only select from LTS (long-term support) versions
[178.62.105.236] out:     --lts=<LTS name>                        When installing, only select from versions for a specific LTS line
[178.62.105.236] out:   nvm uninstall <version>                   Uninstall a version
[178.62.105.236] out:   nvm uninstall --lts                       Uninstall using automatic LTS (long-term support) alias `lts/*`, if available.
[178.62.105.236] out:   nvm uninstall --lts=<LTS name>            Uninstall using automatic alias for provided LTS line, if available.
[178.62.105.236] out:   nvm use [--silent] <version>              Modify PATH to use <version>. Uses .nvmrc if available
[178.62.105.236] out:     --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
[178.62.105.236] out:     --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
[178.62.105.236] out:   nvm exec [--silent] <version> [<command>] Run <command> on <version>. Uses .nvmrc if available
[178.62.105.236] out:     --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
[178.62.105.236] out:     --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
[178.62.105.236] out:   nvm run [--silent] <version> [<args>]     Run `node` on <version> with <args> as arguments. Uses .nvmrc if available
[178.62.105.236] out:     --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
[178.62.105.236] out:     --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
[178.62.105.236] out:   nvm current                               Display currently activated version
[178.62.105.236] out:   nvm ls                                    List installed versions
[178.62.105.236] out:   nvm ls <version>                          List versions matching a given <version>
[178.62.105.236] out:   nvm ls-remote                             List remote versions available for install
[178.62.105.236] out:     --lts                                   When listing, only show LTS (long-term support) versions
[178.62.105.236] out:   nvm ls-remote <version>                   List remote versions available for install, matching a given <version>
[178.62.105.236] out:     --lts                                   When listing, only show LTS (long-term support) versions
[178.62.105.236] out:     --lts=<LTS name>                        When listing, only show versions for a specific LTS line
[178.62.105.236] out:   nvm version <version>                     Resolve the given description to a single local version
[178.62.105.236] out:   nvm version-remote <version>              Resolve the given description to a single remote version
[178.62.105.236] out:     --lts                                   When listing, only select from LTS (long-term support) versions
[178.62.105.236] out:     --lts=<LTS name>                        When listing, only select from versions for a specific LTS line
[178.62.105.236] out:   nvm deactivate                            Undo effects of `nvm` on current shell
[178.62.105.236] out:   nvm alias [<pattern>]                     Show all aliases beginning with <pattern>
[178.62.105.236] out:   nvm alias <name> <version>                Set an alias named <name> pointing to <version>
[178.62.105.236] out:   nvm unalias <name>                        Deletes the alias named <name>
[178.62.105.236] out:   nvm reinstall-packages <version>          Reinstall global `npm` packages contained in <version> to current version
[178.62.105.236] out:   nvm unload                                Unload `nvm` from shell
[178.62.105.236] out:   nvm which [<version>]                     Display path to installed node version. Uses .nvmrc if available
[178.62.105.236] out:   nvm cache dir                             Display path to the cache directory for nvm
[178.62.105.236] out:   nvm cache clear                           Empty cache directory for nvm
[178.62.105.236] out:
[178.62.105.236] out: Example:
[178.62.105.236] out:   nvm install v0.10.32                  Install a specific version number
[178.62.105.236] out:   nvm use 0.10                          Use the latest available 0.10.x release
[178.62.105.236] out:   nvm run 0.10.32 app.js                Run app.js using node v0.10.32
[178.62.105.236] out:   nvm exec 0.10.32 node app.js          Run `node app.js` with the PATH pointing to node v0.10.32
[178.62.105.236] out:   nvm alias default 0.10.32             Set default node version on a shell
[178.62.105.236] out:
[178.62.105.236] out: Note:
[178.62.105.236] out:   to remove, delete, or uninstall nvm - just remove the `$NVM_DIR` folder (usually `~/.nvm`)
[178.62.105.236] out:
[178.62.105.236] out:


Fatal error: sudo() received nonzero return code 127 while executing!

Requested: cd /var/www/quotient&& . ~quotient/.nvm/nvm.sh&& nvm install&& npm install -g yarn&& yarn&& yarn run build
Executed: sudo -S -p 'sudo password:' -H  -u "quotient"  /bin/bash -l -c "cd /var/www/quotient&& . ~quotient/.nvm/nvm.sh&& nvm install&& npm install -g yarn&& yarn&& yarn run build"

Aborting.
```

### LINKED ISSUE #58: Certbot cron job installed by the deploy script is incorrect
Currently:

```
56 3 * * * certbot renew --pre-hook "service nginx stop" --post-hook "service nginx start"
```

I spotted this in syslog while trying to work out why certs were not renewing:

```
Error: bad username; while reading /etc/cron.d/certbot
(*system*certbot) ERROR (Syntax error, this crontab file will be ignored)
```

As this is a system-wide crontab living in `cron.d` cron requires that you supply it a username (right before the command to run) - so it reads this as "run the command 'renew' under the user "certbot". So SSL certificates do not get auto-renewed.

### LINKED ISSUE #60: pushmedia having problems due to permissions
You SSH in and rsync as `deploy`, but the `_media` files and folders need to be owned by the project user, this causes problems.

### PR DESCRIPTION
**Pull request by danielsamuels**
Resolves #18
Resolves #52
Resolves #54
Resolves #55
Resolves #57 
Resolves #58
Resolves #60

Removes support for < Python 3.6.

### CODE REVIEW

File: server_management/management/commands/update.py [hunk 1]
```diff
@@ -208,85 +40,95 @@ def handle(self, *args, **options):
...
+                for line in requirements.split('\n'):
+                    if line.startswith('django-compressor'):
+                        compressor = True
+                    if line.startswith('django-watson'):
+                        watson = True
 
-                    if not compressor:
-                        sudo('python manage.py compileassets')
+                if not compressor:
+                    sudo('python manage.py compileassets')
```
Comment (lewiscollard) [inline server_management/management/commands/update.py]: Is this condition ever going to be reached? I am not sure if any of our `server-management`-using projects use `django-optimizations`, and almost certainly not our Py3 projects (bearing in mind py2 support was removed in this PR).

Comment (danielsamuels) [reply server_management/management/commands/update.py]: That's a fair point. Can probably be removed.

File: server_management/management/commands/deploy.py [hunk 2]
```diff
@@ -290,24 +315,22 @@ def validate_choice(choice):
...
+                'command': f'{pip_command} install -U pip'
             },
             {
                 'title': 'Install virtualenv',
-                'command': '{} install virtualenv'.format(pip_command),
+                'command': f'{pip_command} install virtualenv',
+            },
+            {
+                'title': 'Install supervisor',
+                'command': f'pip2 install supervisor',
```
Comment (dan-gamble) [inline server_management/management/commands/deploy.py:326]: Should this be `pip3`?

File: server_management/management/commands/deploy.py [hunk 3]
```diff
@@ -711,18 +713,24 @@ def validate_choice(choice):
...
+            },
+            {
+                'title': 'Ensure that the staging Nginx config is enabled',
+                'command': 'ln -s /etc/nginx/sites-available/{project}_staging /etc/nginx/sites-enabled/{project}_staging'.format(
                     project=project_folder,
                 ),
             },
             {
                 'title': 'Run certbot',
                 'command': 'certbot certonly --standalone -n --agree-tos --email developers@onespacemedia.com '
```
Comment (dan-gamble) [inline server_management/management/commands/deploy.py:729]: Should this be an `f` string to conform with the rest of the projects strings?

File: server_management/management/commands/deploy.py [hunk 4]
```diff
@@ -711,18 +713,24 @@ def validate_choice(choice):
                 'command': 'rm /etc/nginx/sites-enabled/default',
             },
             {
-                'title': 'Ensure that the application site is enabled',
-                'command': 'ln -s /etc/nginx/sites-available/{project} /etc/nginx/sites-enabled/{project}'.format(
+                'title': 'Ensure that the production Nginx config is enabled',
+                'command': 'ln -s /etc/nginx/sites-available/{project}_production /etc/nginx/sites-enabled/{project}_production'.format(
```
Comment (dan-gamble) [inline server_management/management/commands/deploy.py:717]: Should this be an `f` string to conform with the rest of the projects strings?

File: server_management/management/commands/deploy.py [hunk 5]
```diff
@@ -819,18 +846,17 @@ def validate_choice(choice):
                 },
                 {
                     'title': 'Ensure static folder exists in project',
-                    'command': 'if [ ! -d "/var/www/{project}/{project}/static/" ]; then mkdir /var/www/{project}/{'
-                               'project}/static/; fi'.format(
+                    'command': 'if [ ! -d "/var/www/{project}/{project}/static/" ]; then mkdir /var/www/{project}/{project}/static/; fi'.format(
```
Comment (dan-gamble) [inline server_management/management/commands/deploy.py:849]: Should this be an `f` string to conform with the rest of the projects strings?

File: server_management/management/commands/deploy.py [hunk 6]
```diff
@@ -819,18 +846,17 @@ def validate_choice(choice):
...
                         project=project_folder,
                     ),
                 },
                 {
                     'title': 'Collect static files',
                     'command': '/var/www/{project}/.venv/bin/python /var/www/{project}/manage.py collectstatic '
                                '--noinput --link --settings={project}.settings.{settings}'.format(
```
Comment (dan-gamble) [inline server_management/management/commands/deploy.py:856]: Should this be an `f` string to conform with the rest of the projects strings?

**Comment by dan-gamble** [comment]
Adding to them comments is there any chance we could switch the quiet behaviour so it's loud by default then quieten it where you originally wanted it to be quiet for? I find it quite frustrating running things and getting no feedback as to what's happening and i think at the moment `--debug` doesn't work as intended.

File: server_management/management/commands/deploy.py (hunk 2 above)
Comment (danielsamuels) [reply server_management/management/commands/deploy.py:326]: No, for some reason it doesn't work with pip3. I can't remember why it was.

**Comment by danielsamuels** [comment]
I fixed `--debug` in this PR to work as it should, it was previously broken. For your f-string comments, in situations where the line-length would be greatly increased by using f-string (usually in cases where the same variable is being repeated), I kept the format version.

**Comment by dan-gamble** [comment]
Fair 👍 . I'll just get used to running `--debug`. I've just noticed it's something i do on each project in my `venv` i edit the `server-management` to be verbose by default.

File: server_management/management/commands/update.py (hunk 1 above)
Comment (danielsamuels) [reply server_management/management/commands/update.py]: Actually, this isn't `django-optimizations`, it's `django-compressor` which we definitely still use. Not sure what `compileassets` does, but it was probably added for a reason.

Comment (lewiscollard) [reply server_management/management/commands/update.py]: Are you sure? I was going by [this source file](https://pypkg.com/pypi/django-optimizations/f/src/optimizations/management/commands/compileassets.py) in `django-optimizations` and `manage.py compileassets` not being a known command in a stock project.

Comment (danielsamuels) [reply server_management/management/commands/update.py]: Apparently I can't read the word 'not'

Comment (lewiscollard) [reply server_management/management/commands/update.py]: Which word? :D

**Comment by danielsamuels** [comment]
I'm pretty happy with this PR now. Are we good to merge?

**Comment by dan-gamble** [comment]
If you're happy, i'm happy 👍

**Comment by lewiscollard** [comment]
I am very happy with this. Great work.

**Review by lewiscollard — APPROVED** [review]

**Comment by danielsamuels** [comment]
Awesome, merged! Thanks for the feedback.

**Pull request merged**